content/articles/flow-and-typescript-2.md

```diff
@@ -0,0 +1,112 @@
+}
+```
+When you are starting the conversion, you probably want to remove the `"noImplicitAny": true,` line since it would error if you have any missing type annotations, which is going to be the case pretty much everywhere.
+
+You then need to rename all your `.js` files to `.ts` and `.tsx`. The `.txs` extension is needed if the file contains JSX.
+
+## Some gotchas and examples
+While TypeScript intends to support most of ES6 and ES7 features, you will need to change some bits of your code. Here are the main ones I ran into and some cool TypeScript features.
+
+### Object destructuring assignment
+It doesn't work in TypeScript ([playground](http://www.typescriptlang.org/Playground#src=var%20printer%20%3D%20function%28obj%3A%20Object%29%20{%0A%09console.log%28obj%29%3B%0A}%0A%0Aconst%20hey%20%3D%20{1%3A%20null%2C%202%3A%20null}%3B%0A%0Aprinter%28{%0A%09lol%3A%20null%2C%0A%09...hey%0A}%29%3B)).
+Let's look at the `combineReducers` from redux for example:
+
+```js
+const reducer = combineReducers({
+  router: routerStateReducer,
+  ...reducers,
+});
+// will need to be changed to
+const reducer = combineReducers({
+  router: routerStateReducer,
+  lists: reducers.lists,
+  cards: reducers.cards
+});
+```
+
+### Importing default
+Importing the default from a ES6 module is different from the Babel one. For example in the case of React:
+
+```ts
+// will not work in typescript
+import React from "react";
+// will work in typescript
+import * as React from "react";
+```
+
+### Adding 3rd party definitions
+Adding type definitions for your own app is great but if you are using 3rd party libraries without definitions for them you don't gain much confidence.
+
+Thankfully we have [DefinitelyTyped](https://github.com/borisyankov/DefinitelyTyped) to save us. It's a git repo containing definitions for other a thousand JavaScript libraries.
+There is a tool called [tsd](http://definitelytyped.org/tsd/) to manage definitions but frankly, I don't use so many libraries that it's worth the hassle over manually copying files.
+
+
+### React PropTypes
+Now that you have real types, you can delete all of the React PropTypes and add interfaces for your components props and state like in the following example:
+
+```ts
+interface IAddFormProps {
+  placeholder: string;
+  callback: (value: string) => void;
+}
+
+export class AddForm extends React.Component<IAddFormProps, {}>{...}
+```
+Those props will be checked at compile time rather than runtime and will fail if props are missing or have the wrong type.
+
+
+## Thoughts
+TypeScript and React are working really well together!
+
+The only issues we have encountered so far (and would love help to resolve!) are:
+
+- webpack building the previous version: adding an empty line and saving a file re-triggers a compilation but can be frustrating
+- no string enums: we want to keep human readable enum members but TypeScript only supports int enums
```

Comment:
Also needs a lot more type annotations. ts really isn't very good at inferences, I'm basically writing java at this point :)